devicetree: gen_defines: Generate lists of children node identifiers

de-nordic opened the pull request:
For each node, a list of identifiers of children nodes is generated
in form of:
DT_S_<path_to_node>_CHILD_IDENTIFIERS <id0>[, <id1>[, ...]]

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

Purpose of this commit is to generate list of identifiers from dts nodes, to be able to more automatically traverse dts tree.
For example useful to process list of flash partitions.

de-nordic committed:
devicetree: gen_defines: Generate list of nodes' children

For each node, a list of identifiers of children nodes is generated
in form of:
DT_S_<path_to_node>_CHILD_IDENTIFIERS <id0>[, <id1>[, ...]]

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

mbolivar-nordic commented on include/devicetree.h:
```diff
@@ -145,6 +145,15 @@
 
+/**
+ * @brief returns list of child node identifiers.
+ * List contains only praimary enabled child nodes (no aliases).
```
Since DT_PARENT() returns the parent regardless of if it's enabled, I think this macro should do the same. The user can always skip generating code for a particular child if it's not enabled using something like `COND_CODE_1(DT_HAS_NODE(child), (DO_SOMETHING_WITH(child),), ())`

de-nordic replied:
OK

mbolivar-nordic replied:
Just checking -- this will be part of a later update, right? It doesn't seem done yet when I checked the output. For example, when building hello_world for nRF52840-DK, I see:

```
/* Node children (/) identifiers: */
#define DT_N_CHILD_IDENTIFIERS sw_pwm, arduino_header, arduino_adc
```

But `/` has more children not in that list (like `soc`, `leds`, etc.)

de-nordic replied:
Originally I have been going after the labels, the things ending with ':', `soc`, `leds` does not end with `:` . I think these should be on different lists then as someone may use own labels , eg `alarm_led: led_0`

mbolivar-nordic replied:
Oh, I missed that, I'm sorry. Hrm, this is new, then. We've normally just operated on nodes themselves. Node labels were just a way to get to the nodes.

This starts to feel tricky because a node can have zero, one, or many nodes, so this is potentially looping over child nodes multiple times. I see why the flash partitions work wants to operate on node labels so you can do things like create an enum with 'boot_partition' in the identifier. But I think this solution probably ought to work even with nodes that do not have labels, since their presence is not guaranteed.

de-nordic replied:
@mbolivar-nordic  OK, I fix it to generate list of all children by the node identifier, not node label. It will not help me with what I do, but will be useful  anyway, so no reason to waste the effort

mbolivar-nordic replied:
> It will not help me with what I do, but will be useful anyway, so no reason to waste the effort

How come?

I guess I was figuring something like this (completely untested):

```
#define FLASH_AREA(partition_node_identifier) flash_area_##partition_node_identifier

#define FLASH_MEMORY DT_CHOSEN(zephyr_flash)
#define FLASH_PARTITIONS DT_CHILD(FLASH_MEMORY, partitions)

#define MAKE_FLASH_ENUM(partition_node) FLASH_AREA(partition_node),
enum my_partitions {
    SOME_UTIL_FOR_EACH_MACRO_TBD(MAKE_FLASH_ENUM, DT_CHILDREN(FLASH_PARTITIONS))
};
```

Now, if you want to know the flash enumerator for the boot partition, you can use something like `FLASH_AREA(DT_NODELABEL(boot_partition))`.

What am I missing here?

de-nordic replied:
>```
> #define FLASH_AREA(partition_node_identifier) flash_area_##partition_node_identifier
> ...
> FLASH_AREA(DT_NODELABEL(boot_partition))
> ```

Invocations like that do not work, because C preprocessor, when evaluating macro, first does parameter substitution and processes # and ## operators, before starting with replacing macro names with definitions. This means that above will stop at:
```
flash_area_DT_NODELABEL(boot_partition)
```

mbolivar-nordic replied:
The usual trick to resolve that is to do another layer of expansion, like with 63c936df7d1e6a325717e4554c301bf04c3d4983

de-nordic replied:
Thanks, done it.

mbolivar-nordic commented on include/devicetree.h:
```diff
@@ -145,6 +145,15 @@
+ * @param path parent node path
+ * @return list of child node identifiers under path.
+ */
+#define DT_CHILD_NODELABELS(path) DT_CAT(path, _CHILD_NODELABELS)
```
I think DT_CHILDREN() would be a better name. These are not node labels, they are node identifiers, and we don't want to mix those two concepts up in the API names.

de-nordic replied:
OK

mbolivar-nordic replied:
It would also be nice to see this placed after DT_PARENT() and DT_CHILD() in devicetree.h.

de-nordic replied:
ok

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -68,10 +68,13 @@ def main():
         write_global_compat_info(edt)
 
 
+
+
```
```suggestion
```

de-nordic replied:
OK

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -68,10 +68,13 @@ def main():
             write_special_props(node)
             write_vanilla_props(node)
 
+        write_child_identifiers(edt)
```
I think this should go after the parent print above, to write it for all nodes.

de-nordic replied:
Done

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -359,6 +362,29 @@ def write_compatibles(node):
 
+def write_child_identifiers(edt):
+    # Writes macro that lists identifiers of all children per node.
+    # Only enabled nodes are processed and identifiers for enabled child
```
This also should change

de-nordic replied:
Done

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -359,6 +362,29 @@ def write_compatibles(node):
+            # Out list of labels only if non empty and remove trailing
+            # ', ' prior.
+            if labels:
+                out_dt_define(node.z_path_id + '_CHILD_NODELABELS',
```
requires macros.bnf update

de-nordic replied:
Done

mbolivar-nordic replied:
Looks like the change from _CHILD_NODELABELS to _CHILDREN was only done for the API macro name -- could you do it for this too?

de-nordic replied:
ok

nvlsianpu commented on scripts/dts/gen_defines.py:
```diff
@@ -359,6 +361,24 @@ def write_compatibles(node):
+        # Out list of labels only if non empty and remove trailing
+        # ', ' prior.
+        if labels:
+            out_dt_define(node.z_path_id + '_CHILD_IDENTIFIERS',
```
shell we have a comment generated like for the parrent macro:

```suggestion
            out_comment(f"Node children ({node.path}) identifiers:")
            out_dt_define(node.z_path_id + '_CHILD_IDENTIFIERS',
```

nvlsianpu replied:
output checked, looks promising
good even without below improvement

de-nordic replied:
Done

galak commented on include/devicetree.h:
```diff
@@ -145,6 +145,15 @@
 
+/**
+ * @brief returns list of child node identifiers.
+ * List contains only praimary child nodes (no aliases).
```
typo: praimary, primary.

de-nordic replied:
Fixed.

de-nordic committed:
devicetree: Add DT_FOR_EACH_CHILD macro

The macro iterates through the list of child nodes and invokes provided
macro for each node.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
https://github.com/zephyrproject-rtos/zephyr/pull/24848 has been added on top, as there is currently no reason to review them separately.

de-nordic commented on include/devicetree.h:
```diff
@@ -1503,6 +1512,79 @@
+ * of first instance of "fixed-partitions" compatible.
+ *
+ */
+#define DT_FOR_EACH_CHILD(parent, fun) DT_FOR_EACH_CHILD_(parent, fun)
```
There is a reason to do this here instead of using FOR_EACH from util, you can do:
`DT_FOR_EACH_CHILD(something, FOR_EACH(sth else))`
but you can not do:
`FOR_EACH(something, FOR_EACH(sth else)`

wentongwu reviewed (approved):
not familiar with DT, but wandering is there any plan to implement the Marco to get all the parent nodes given a node? or we already have it? This should benefit power management.

de-nordic commented:
> 
> 
> not familiar with DT, but wandering is there any plan to implement the Marco to get all the parent nodes given a node? or we already have it? This should benefit power management.

If you want to traverse the DT path back, you can use the DT_PARENT() macro as defined in includes/devicetree.h. It will give you path to the parent of given path to node.

Check the include/generated/devicetree_unfixed.h file in your build directory to get more practical view on how are the PARENT references generated.

galak commented:
> not familiar with DT, but wandering is there any plan to implement the Marco to get all the parent nodes given a node? or we already have it? This should benefit power management.

Don't you really want the dependency graph, not just parents.  I know @mbolivar-nordic and @pabigot are thinking about this.  It seems all tied up with how we move forward with struct device as well and work @tbursztyka has been looking at.

pabigot commented:
We do indeed have the dependency graph, and can produce the set of nodes that a given node depends on, but we need a simple handle to identify the corresponding devices and init sections.  The immediate blocker for that is the ability to represent devices and init routines by integral indexes, which is held up on the device structure rework.

wentongwu commented:
> Don't you really want the dependency graph, not just parents.

@galak @zephyrbot I think there are two kinds of use cases for the dependency graph on power management side, the first one is to power on/off all the device nodes following parent-children dependency, maybe a list where every node's parent is ahead of the node self can satisfy this case; the other one is when no traffic on one node and all of it's children are powered off, we can power off this node and then look at his parents,..., some like BFS of graph, not sure how the graph will be structured and represented, but seems easily getting children set and parent set can make the second case work.

wentongwu commented:
> > not familiar with DT, but wandering is there any plan to implement the Marco to get all the parent nodes given a node? or we already have it? This should benefit power management.
> 
> If you want to traverse the DT path back, you can use the DT_PARENT() macro as defined in includes/devicetree.h. It will give you path to the parent of given path to node.
> 
> Check the include/generated/devicetree_unfixed.h file in your build directory to get more practical view on how are the PARENT references generated.

@de-nordic so if want to get a node's parent, first I should get node's path identifier, and then get the parent identifier with macro `NODE_PATH_IDENTIFIER ## _PARENT` (e.g. DT_N_S_cpus and DT_N_S_cpus_PARENT). But at runtime, we are doing things with `struct device` , how do we get parent with `struct device`?

mbolivar-nordic commented:
>  how do we get parent with `struct device`?

If you have `PARENT_ID`, a node identifier for the node's parent, *and* you are sure there is a struct device for that node, you can just use `device_get_binding(DT_LABEL(PARENT_ID))` in the same way you would any other time you wanted a struct device.

https://docs.zephyrproject.org/latest/guides/dts/howtos.html#get-a-struct-device-from-a-devicetree-node

Node identifiers are all the same; it doesn't matter how you get one.

wentongwu commented:
> > how do we get parent with `struct device`?
> 
> If you have `PARENT_ID`, a node identifier for the node's parent, _and_ you are sure there is a struct device for that node, you can just use `device_get_binding(DT_LABEL(PARENT_ID))` in the same way you would any other time you wanted a struct device.
> 
> [docs.zephyrproject.org/latest/guides/dts/howtos.html#get-a-struct-device-from-a-devicetree-node](https://docs.zephyrproject.org/latest/guides/dts/howtos.html#get-a-struct-device-from-a-devicetree-node)
> 
> Node identifiers are all the same; it doesn't matter how you get one.

@mbolivar-nordic thanks for the detail, not familiar with DTS so maybe I didn't describe the question clearly and yes, we can get the `struct device` by `device_get_binding` if we have node identifier. But at runtime, we are playing with `struct device`, in order to get the parent first we should know the child's node identifier following @de-nordic , but do we already have the method to get the node identifier for the given `struct device`? or we have other ways handling device dependency?

pabigot commented:
> ... do we already have the method to get the node identifier for the given `struct device`? or we have other ways handling device dependency?

No.  That's what we're discussing at https://github.com/zephyrproject-rtos/zephyr/issues/22941#issuecomment-636848156.  I expect a solution in the 2.4.0 development cycle.
